### PR TITLE
feat: I guess it's TypeScript

### LINKED ISSUE #958: RAM rises on v1.19.0
Windows 11 Pro 22H2/Core i5-8350U/RAM 12GB

I tried to download a playlist of 103 songs. The RAM useage before download was 3.7GB (31%), but it gradually rose to around 90% after 7 mins or so as download proceeded, and download just stopped nearly at the end after 101 songs. The download process was still running, and the RAM useage remained around 90% untill I terminated the app. It seems like a typical memory leak situation. Any thoughts? 

If I do the same download on Core i5-1035G1/RAM 24GB, its RAM useage went up to 14GB (59%) and stayed there after 101 songs untill I killed the app. It seems it was trying to download or encode the rest two songs, and some process internally hangs up?

Thx!

### LINKED ISSUE #962: Migration to Typescript
Will be doing this
Please assign me to this.

### LINKED ISSUE #978: Remove `md5` dependency
https://github.com/th-ch/youtube-music/blob/3a822f611af13f31d74f5c4636cc73534548aa56/plugins/last-fm/back.js#L2
https://github.com/th-ch/youtube-music/blob/3a822f611af13f31d74f5c4636cc73534548aa56/plugins/last-fm/back.js#L40

can be replaced with `node:crypto`'s `createHash`

Example
```js
const crypto = require("node:crypto")

sig = crypto.createHash("md5").update(sig, "utf-8").digest("hex")
```

### LINKED ISSUE #1147: 1.20.0 Load Time
Let me know how I can help, but 1.20.0 takes a really long time to load vs 1.19.0 and earlier.  For now I have downgraded and things are fine again.

In all cases I am using the installed/setup version.

I tried upgrading and fresh installs.
I did all the normal, reboot my PC type little things.

Windows 11 Pro
32GB RAM
2TB M.2 SSD

1.20.0 does eventually load, but 1.19.0 (and earlier) takes about 1.6 seconds to load from shortcut click.  1.20.0 takes about 45 seconds to load up and even then is not really ready to be used for another 15-30 seconds.

### LINKED ISSUE #1158: Memory leak after continuous useage
The mem leak doesn't take long to show; give it a couple hours to see the RAM start rising. Leave music playing all day or overnight and you'll come back to a frozen computer (or YT music will crash).

This is an ongoing issue with native YT music, actually. Can we, perhaps, _auto-restart the app after x amount of time_ as a workaround?

Related: 
* https://github.com/th-ch/youtube-music/issues/797
* https://github.com/th-ch/youtube-music/issues/322
* https://github.com/th-ch/youtube-music/issues/295
* https://github.com/th-ch/youtube-music/issues/958
* https://www.reddit.com/r/YoutubeMusic/comments/vhvark/web_app_still_has_a_memory_leak/ 
* https://support.google.com/youtubemusic/thread/69139975/web-ytm-leaks-memory-and-forces-me-to-reload-the-tab-every-single-day?hl=en

### LINKED ISSUE #1162: New version Lag (1.20)
Application lags on the following:

-Opening of application.
-Song selections. 
-Buffering through most of songs notices up to 10 songs. I verified it is not my Bandwidth as another user has the same problem.
- Any control function used.

Remedies tried:
-Reduced quality of music videos. (does not seem to affect the issue)
-Task kill and restart. (works 95% of the time)
-Reinstall of application.(seemed to fix the issue for first use after install. closed application  and opened issue persists)

Hopefully this is useful information just wanted to add to this project in any way since I am an avid user :)

### LINKED ISSUE #1172: Version 1.20 ignores downloader "ffmpegArgs"
The  version 1.20 on windows 10 x64 ignores any "ffmpegArgs" and saves to default mp3 124 kbs. 

			"ffmpegArgs": [
				"-b:a",
				"384k",
				"-ar",
				"44100"
			],

Installing the version 1.19 same arguments work as intended and file is saved to mp3 384k 44100

Note: it was a fresh win 10 install and 1.19 was installed after uninstalling the 1.20. (because 1.20 ignored the args).
Anyone else please check also to see if this is only a quirk of my machine or is realy a issue related to 1.20.
Thank you!

### LINKED ISSUE #1173: Discord Rich presence won't work properly or be shown properly while while playing a song etc.
Extra Information:
Yt music client .deb packages
Discord Client .deb packages

### LINKED ISSUE #1187: Adblock mode causes memory issues; "In-Player" fixes memory leakage
## The Problem

Recent memory accumulation events have been reported. One solution working for me and a majority of users was to change the Adblock plugin mode to "In-Player" from "with blocklists"

image

## My Findings

After running multi-hour tests with the Windows Performance Monitor, RAM usage seems to be stable (fig 1) using "in-player", but sees an accumulation of memory over time (fig 2) using "with blocklists". These findings correspond with the smooth experience while using "in-player" but laggy experience (after a period of time) using "with blocklists".

### fig 1.

Performance Monitor 6_30_2023 04_02_24 PM

### fig 2.

Performance Monitor 6_30_2023 04_02_28 PM

Related #1158 , #1147 , #1162

### LINKED ISSUE #1188: Donloading in MP3 crashes the application
Hey,
after a while I've come to use the downloader plugin once more, only to find the whole application crashing when downloading anything in MP3.
I'm not completely certain - but I  think to have seen it working under 1.20 already on my windows 11 machine. So theoretically nothing should have changed, or are plugins updated silently in the background, separately from the main application? My OS environment should also be pretty much unchanged, execept for some small cumulative Windows updates maybe.
Downloading in opus works flawlessly - and faster, without the conversion to MP3 - but I really like having my id3 data. Thank you for adding that btw.

Greetings
TJW65

### LINKED ISSUE #1208: Fixes for 1.20 issues in one place to help out folks.
If you're experiencing severe lag or playback issues in the app, switch the adblocker mode from "blocklists" to "in player", and if that doesn't fix it, try turning off ad block completely if you already have a premium account or don't care about ads.

If you don't see an image in your player, turn off the bypass age restriction feature.

Some users are also reporting that they fixed longterm memory leaks by turning off genius lyrics.

There are also currently more issues I haven't seen a fix for, such as the image resolution for music cover art being very low or missaligned, and the new version of the youtube music app just, lacks an actual back button?

### LINKED ISSUE #1212: extreme lag in the client
the ui is extremely laggy, music plays but when it lags sometimes it just skips songs, its started randomly

### LINKED ISSUE #1226: app is unusable recent version !!
bigger leaks than when Kakhovka dam was blown. Windows is dying from memory overload

no fix comes ??

### LINKED ISSUE #1227: Taking Too Much CPU Usage!
Hey can you please let me know how can I optimize the app so It takes less CPU usage I'm a gamer and I love listening to music in background but It's taking some of my FPS, I'm using default settings but is there anything I can Enable/Disable to make it only use like 100 MB of memory If possible.

### LINKED ISSUE #1240: downloading playlists (mp3) does not work/crashes app
Hi, im trying to download playlists, im on windows and tried it with both versions 1.19 and 1.20.
When using 1.19, the app does not crash, the folder is created but no songs are downloaded, it just shows the animaiton in the taskbar, and the message still says: "Downloading" when downloading a playlist. When i try to download a song it shows "Downloading..." but does not download the song. When using 1.20, after the first song of the playlist the app crashes completely. Also tried it with different playlists, same behaviour.

### LINKED ISSUE #1247: Taskbar Media Controls Not Working
Enabled Taskbar Media Controls but the controls aren't showing in the app thumbnail when hovering over the app in the taskbar?

### LINKED ISSUE #1249: Experiencing lag after playing a lot of music
I've been experiencing this after i played lots of music like there's a delay whenever i click something and the music stop for about a second then play again but at first it runs so smooth. restarting the app fix it but its so hassle to do it all the time.
i never experience it in youtube music in browser even when im playing and do lots of task, only in this app.
my unit is ryzen 5 5600g with 16gb of ram so i think my pc is not the problem.

hope will solve this problem thank you dev.

### PR DESCRIPTION
**Pull request by JellyBrick**
- resolves #962 (diff: 3e3fdb3...451d337)
- resolves #958 
- resolves #1240 
- resolves #1188 
- resolves #1172
- resolves #978 
- resolves #1247
- resolves #1187 (resolves #1208, resolves #1212, resolves #1226, resolves #1227, resolves #1147, resolves #1158, resolves #1162, resolves #1249)
 this issue is related to https://github.com/ghostery/adblocker/issues/3420
- Resolved an issue where `pause` events are not handled (resolves #1173)
- Migration from CJS to ESM (use "CommonJS" mode, as Electron does not support ESM)
- Use Eslint + TypeScript parser
- Use 2-space indent (The original code was a mix of 4-space, 1-tab, and 2-space😅)
- Compile-time configuration validator
image
image
- Use npm instead of yarn legacy (v1)
- Update dependencies version
- Pin dependencies version
- Remove `node-fetch`. use `electron.net.fetch` instead. ([it support `automatic management of system proxy configuration`, and etc.](https://www.electronjs.org/docs/latest/api/net))
- Remove unused dependencies. (`browser-id3-writer`, `md5`)
- Create type definitions for JavaScript libraries (e.g. `custom-electron-prompt`, `custom-electron-titlebar` (original d.ts is weird), `keyboardevent-from-electron-accelerator`, `keyboardevents-areequal`, `mpris-service`, `vudio`, `butterchurn`)
- Create type definitions for YouTube Internal API, SponsorBlock API, `window.navigation` API, CustomEvent

### Question

- Why should the downloader work synchronously? It takes too long when download 100+ songs.

## Image

image

It works with plugins! (100% drop-in replacement of original version)

### CODE REVIEW

**Comment by ArjixWasTaken** [comment]
I guess this is more true to the original project than my crappy re-write.

Great work!

File: types/youtube-player.ts [hunk 1]
```diff
@@ -0,0 +1,189 @@
+// TODO: fully type definitions for youtube-player
```
Comment (ArjixWasTaken) [inline types/youtube-player.ts:1]: Feel free to use my partial typings.

https://github.com/ArjixWasTaken/youtube-music-ts/blob/master/src/shared/types/PlayerAPI.ts

**Review by ArjixWasTaken — COMMENTED** [review]
Btw, instead of npm you should really use pnpm.

File: types/youtube-player.ts (hunk 1 above)
Comment (ArjixWasTaken) [reply types/youtube-player.ts:1]: The changes are so many that the GitHub UI is struggling when viewing them 🤣

**Comment by JellyBrick** [comment]
> Btw, instead of npm you should really use pnpm.

`pnpm` causes an issue when used with an `electron-builder`.

https://github.com/electron-userland/electron-builder/issues/6289

**Comment by JellyBrick** [comment]
The source of the adblocker's memory leak is in the library, not the application's code.

https://github.com/ghostery/adblocker/issues/3420

**Comment by JellyBrick** [comment]
I don't know why Snyk is failing.

**Comment by SomeAspy** [comment]
This is amazing work, I hope to see this get merged

**Comment by JellyBrick** [comment]
image

Here's why SNYK fails. (we can just ignore it)

**Comment by ArjixWasTaken** [comment]
> Here's why SNYK fails. (we can just ignore it)

did they grand you access to the snyk logs?

**Comment by JellyBrick** [comment]
> > Here's why SNYK fails. (we can just ignore it)
> 
> did they grand you access to the snyk logs?

I ran Snyk separately on my own fork.

**Comment by ArjixWasTaken** [comment]
It would be so much easier to review if the entire codebase wasn't reformatted as well.

Is it possible for you to apply your changes w/o formatting the code? Or are there too many changes for you to do that?

**Comment by ArjixWasTaken** [comment]
Maybe [git mv](https://git-scm.com/docs/git-mv) can be used to achieve this.

Although that would require someone to reapply (manually) all the changes you did.
Which might not be that bad, I'll see if I can do this myself on a small subset of commits.

**Comment by JellyBrick** [comment]
@ArjixWasTaken 

GitHub appears to treat "huge changed code" as a separate file.

# Small change

image

# Huge change


image

**Comment by JellyBrick** [comment]
To see the differences more easily, use this link: https://github.com/th-ch/youtube-music/compare/3e3fdb3c3f631dbad47caaa5620d876f5641d114...c8b149281b34ae33c28de3602740c3cbbebcc3f4

**Comment by JellyBrick** [comment]
@th-ch Could you review this PR?

**Review by th-ch — APPROVED** [review]
✅ Looks good, thanks for the hard work! Hard to review everything but overall, it looks ok, if any bugs are found they can be fixed with follow up PRs - probably worth a new major version as well considering the size of changes

**Pull request merged**